### ISSUE TITLE
Info on finetune scripts

### ISSUE THREAD

**Issue by gkiri** [OpenBMB/VoxCPM#114]
Thank you for updating finetune support.
before diving in to finetuning , could you please share few details that can help the journey
1. can we finetune on any new languages other than English and Chinese ?
2. How many hours would be optimal for good finetune results ?

**Comment by Ayin1412** [comment]
tested on japanese and korean, the finetuned model is quite good, and learned how to punctuate in the language(mostly imitating the training data).

my training data is about 2hours

**Comment by gkiri** [comment]
Wow that sounds great . Looking forward to training on new language .

**Comment by picolo100** [comment]
> tested on japanese and korean, the finetuned model is quite good, and learned how to punctuate in the language(mostly imitating the training data).
> 
> my training data is about 2hours

how we can do that, i  wanna do it in my language to, could you provide the step by step, i'll thank you with a dono np

**Comment by Ayin1412** [comment]
> > tested on japanese and korean, the finetuned model is quite good, and learned how to punctuate in the language(mostly imitating the training data).
> > my training data is about 2hours
> 
> how we can do that, i wanna do it in my language to, could you provide the step by step, i'll thank you with a dono np

i've just record a video to show the procedures(in P2), (forgive my poor english),[Video Here](https://www.bilibili.com/video/BV121mTBQEoU?p=2)
and here is my blog about how to use the webui, but it's in chinese, you need to translate it [Post Here](https://www.hokma.top/posts/voxcpm/)

**Comment by picolo100** [comment]
> > > tested on japanese and korean, the finetuned model is quite good, and learned how to punctuate in the language(mostly imitating the training data).
> > > my training data is about 2hours
> > 
> > 
> > how we can do that, i wanna do it in my language to, could you provide the step by step, i'll thank you with a dono np
> 
> i've just record a video to show the procedures(in P2), (forgive my poor english),[Video Here](https://www.bilibili.com/video/BV121mTBQEoU?p=2) and here is my blog about how to use the webui, but it's in chinese, you need to translate it [Post Here](https://www.hokma.top/posts/voxcpm/)

im gonna test it when i wake up your are brave! keep you updated on my progressions.

**Comment by picolo100** [comment]
> > > tested on japanese and korean, the finetuned model is quite good, and learned how to punctuate in the language(mostly imitating the training data).
> > > my training data is about 2hours
> > 
> > 
> > how we can do that, i wanna do it in my language to, could you provide the step by step, i'll thank you with a dono np
> 
> i've just record a video to show the procedures(in P2), (forgive my poor english),[Video Here](https://www.bilibili.com/video/BV121mTBQEoU?p=2) and here is my blog about how to use the webui, but it's in chinese, you need to translate it [Post Here](https://www.hokma.top/posts/voxcpm/)

unfortunately i cant correctly follow the video as i can't access to bilibli i don't have yet a chinese phone number therefore the video is 360p resolution and unseeable, could you upload it to another plateform or share it in telegram with me, but i prefer if you make it public as other people can access is to, it's at your ease.

**Comment by progmars** [comment]
I have trained it to speak Latvian. Used about 100h of data from Mozilla Common Voice dataset (needed some preprocessing to remove quotes and empties and entries with broken Unicode symbols) and let it train for 16 hours in the background on my 3090 with power limit. Had to decrease batch size to 4 and correspondingly set grad_accum_steps to 4, but honestly I have no idea what's best, I'm quite new to training.

It became fluent after 20k steps but needed 60k steps to iron out some small quirks. However, the results are wavy - the cosine learning rate affects it I think. The quality of checkpoints goes up and down. If it's bad, it will get loud, metallic and raspy and lose cloned voice characteristics for longer sentences. And Tensorboard graphs were not easy to understand for this model, no classic loss/validation relation, and the loss doesn't seem to go down at all, even when the quality improves, so I was quite in the dark about what I'm actually doing. I'm sure an experienced person would be able to understand it better.

So, to test the quality and decide which checkpoint to keep, I created a script to generate a few samples for every checkpoint and listened to them. Also tested with different cfg values. In my case, I found that cfg=2.7 is a good balance. 3 and above sometimes can get loud, metallic and raspy towards the end.

**Comment by picolo100** [comment]
> I have trained it to speak Latvian. Used about 100h of data from Mozilla Common Voice dataset (needed some preprocessing to remove quotes and empties and entries with broken Unicode symbols) and let it train for 16 hours in the background on my 3090 with power limit. Had to decrease batch size to 4 and correspondingly set grad_accum_steps to 4, but honestly I have no idea what's best, I'm quite new to training.
> 
> It became fluent after 20k steps but needed 60k steps to iron out some small quirks. However, the results are wavy - the cosine learning rate affects it I think. The quality of checkpoints goes up and down. If it's bad, it will get loud, metallic and raspy and lose cloned voice characteristics for longer sentences. And Tensorboard graphs were not easy to understand for this model, no classic loss/validation relation, and the loss doesn't seem to go down at all, even when the quality improves, so I was quite in the dark about what I'm actually doing. I'm sure an experienced person would be able to understand it better.
> 
> So, to test the quality and decide which checkpoint to keep, I created a script to generate a few samples for every checkpoint and listened to them. Also tested with different cfg values. In my case, I found that cfg=2.7 is a good balance. 3 and above sometimes can get loud, metallic and raspy towards the end.

could you please tell me the steps to do it for faster and clear understanding?

**Comment by Ayin1412** [comment]
> > > > tested on japanese and korean, the finetuned model is quite good, and learned how to punctuate in the language(mostly imitating the training data).
> > > > my training data is about 2hours
> > > 
> > > 
> > > how we can do that, i wanna do it in my language to, could you provide the step by step, i'll thank you with a dono np
> > 
> > 
> > i've just record a video to show the procedures(in P2), (forgive my poor english),[Video Here](https://www.bilibili.com/video/BV121mTBQEoU?p=2) and here is my blog about how to use the webui, but it's in chinese, you need to translate it [Post Here](https://www.hokma.top/posts/voxcpm/)
> 
> unfortunately i cant correctly follow the video as i can't access to bilibli i don't have yet a chinese phone number therefore the video is 360p resolution and unseeable, could you upload it to another plateform or share it in telegram with me, but i prefer if you make it public as other people can access is to, it's at your ease.

left your email pls,i'll send it to you

**Comment by progmars** [comment]
> could you please tell me the steps to do it for faster and clear understanding?

It all depends on the quality of dataset. Many open datasets have been made by asking people to volunteer for reading random text fragments, and this, unfortunately, can make the result sound monotone and emotionless, which would be good only for some corporate video dubbing but not conversational bots and adventure stories.

I did not change any training parameters in the default training example, but had to adjust batch size just because of running out of video memory. As I had downloaded the dataset to disk, I had to also adjust its paths and also the load_dataset arguments to support tsv. Even then, it broke because of quotes in the data, so I had to clean up the texts by simply replacing all "" with empty text. Some texts in the dataset were corrupted because of Unicode conversion mistakes that Mozilla had - so I had to remove those entries too, to have cleaner dataset.

The training script spent quite a long time loading and mapping the dataset - just have to be patient and wait, and then training will eventually begin. I could also interrupt it and continue the next day - it can pick up the last checkpoint. 

I also set 
num_iters: 100000
so I could leave it alone and be sure it will continue training.

And, of course, manual evaluation (generating a bunch of files using your trained checkpoints) is the key to avoid overtraining and find the best checkpoint.

**Comment by picolo100** [comment]
> > > > > tested on japanese and korean, the finetuned model is quite good, and learned how to punctuate in the language(mostly imitating the training data).
> > > > > my training data is about 2hours
> > > > 
> > > > 
> > > > how we can do that, i wanna do it in my language to, could you provide the step by step, i'll thank you with a dono np
> > > 
> > > 
> > > i've just record a video to show the procedures(in P2), (forgive my poor english),[Video Here](https://www.bilibili.com/video/BV121mTBQEoU?p=2) and here is my blog about how to use the webui, but it's in chinese, you need to translate it [Post Here](https://www.hokma.top/posts/voxcpm/)
> > 
> > 
> > unfortunately i cant correctly follow the video as i can't access to bilibli i don't have yet a chinese phone number therefore the video is 360p resolution and unseeable, could you upload it to another plateform or share it in telegram with me, but i prefer if you make it public as other people can access is to, it's at your ease.
> 
> left your email pls,i'll send it to you

vestoslipher117@gmail.com

**Comment by picolo100** [comment]
> > > > > tested on japanese and korean, the finetuned model is quite good, and learned how to punctuate in the language(mostly imitating the training data).
> > > > > my training data is about 2hours
> > > > 
> > > > 
> > > > how we can do that, i wanna do it in my language to, could you provide the step by step, i'll thank you with a dono np
> > > 
> > > 
> > > i've just record a video to show the procedures(in P2), (forgive my poor english),[Video Here](https://www.bilibili.com/video/BV121mTBQEoU?p=2) and here is my blog about how to use the webui, but it's in chinese, you need to translate it [Post Here](https://www.hokma.top/posts/voxcpm/)
> > 
> > 
> > unfortunately i cant correctly follow the video as i can't access to bilibli i don't have yet a chinese phone number therefore the video is 360p resolution and unseeable, could you upload it to another plateform or share it in telegram with me, but i prefer if you make it public as other people can access is to, it's at your ease.
> 
> left your email pls,i'll send it to you

u here brother, u didnt send the video please im waiting for the work

**Comment by Ayin1412** [comment]
> > > > > > tested on japanese and korean, the finetuned model is quite good, and learned how to punctuate in the language(mostly imitating the training data).
> > > > > > my training data is about 2hours
> > > > > 
> > > > > 
> > > > > how we can do that, i wanna do it in my language to, could you provide the step by step, i'll thank you with a dono np
> > > > 
> > > > 
> > > > i've just record a video to show the procedures(in P2), (forgive my poor english),[Video Here](https://www.bilibili.com/video/BV121mTBQEoU?p=2) and here is my blog about how to use the webui, but it's in chinese, you need to translate it [Post Here](https://www.hokma.top/posts/voxcpm/)
> > > 
> > > 
> > > unfortunately i cant correctly follow the video as i can't access to bilibli i don't have yet a chinese phone number therefore the video is 360p resolution and unseeable, could you upload it to another plateform or share it in telegram with me, but i prefer if you make it public as other people can access is to, it's at your ease.
> > 
> > 
> > left your email pls,i'll send it to you
> 
> u here brother, u didnt send the video please im waiting for the work

double check pls
maybe in the dustbin

<img width="2345" height="276" alt="Image" src="https://github.com/user-attachments/assets/98c5c190-e4cf-41b7-b75b-7efc59539428" />

**Comment by picolo100** [comment]
> > > > > > > tested on japanese and korean, the finetuned model is quite good, and learned how to punctuate in the language(mostly imitating the training data).
> > > > > > > my training data is about 2hours
> > > > > > 
> > > > > > 
> > > > > > how we can do that, i wanna do it in my language to, could you provide the step by step, i'll thank you with a dono np
> > > > > 
> > > > > 
> > > > > i've just record a video to show the procedures(in P2), (forgive my poor english),[Video Here](https://www.bilibili.com/video/BV121mTBQEoU?p=2) and here is my blog about how to use the webui, but it's in chinese, you need to translate it [Post Here](https://www.hokma.top/posts/voxcpm/)
> > > > 
> > > > 
> > > > unfortunately i cant correctly follow the video as i can't access to bilibli i don't have yet a chinese phone number therefore the video is 360p resolution and unseeable, could you upload it to another plateform or share it in telegram with me, but i prefer if you make it public as other people can access is to, it's at your ease.
> > > 
> > > 
> > > left your email pls,i'll send it to you
> > 
> > 
> > u here brother, u didnt send the video please im waiting for the work
> 
> double check pls maybe in the dustbin
> 
> <img alt="Image" width="2000" height="276" src="https://private-user-images.githubusercontent.com/61609756/528487824-98c5c190-e4cf-41b7-b75b-7efc59539428.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjYxMzAyNDQsIm5iZiI6MTc2NjEyOTk0NCwicGF0aCI6Ii82MTYwOTc1Ni81Mjg0ODc4MjQtOThjNWMxOTAtZTRjZi00MWI3LWI3NWItN2VmYzU5NTM5NDI4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjE5VDA3MzkwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThiMmZhMjhlYWRjYmMwM2I5OGRmYTY1YWJjOWI0YThhYjhjMzE5YWRlZmUwNTg5ZGI4Nzk3OTYwYmEwMDgyYWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.z3OnmdYOk8RqI5mKTAEDTp2dcJ53_mByipuFBTFezgY">

i find it i'll gonna check it and by the way do you have a fast contact where i can connect with you quick for the work?
